# *BEANS*, Bioinformatic envioriment for variant analysis SNPs
## User manual and guide 


## Overview

BEANS is a python pipeline designed for the detection of genomic variants, SNP and Indels, using two recognized methods: [*Freebayes*](http://arxiv.org/abs/1207.3907) and [*GATK*](https://gatk.broadinstitute.org/hc/en-us).

This project represents a joint effort of the DNA and Genomics Laboratory at the National Center for Genetic Resources of INIFAP, aiming to develop accessible and easily executable workflows for users of all backgrounds.

## Required Packages

We highly recommend installing the necessary packages using [Conda](https://www.anaconda.com/download) or [Docker](https://www.docker.com).

For the Freebayes method:

- Python 3 or higher
- [Fastp](https://github.com/OpenGene/fastp)
- [BWA-mem2](https://github.com/bwa-mem2/bwa-mem2)
- [Samtools](https://github.com/samtools/samtools)
- [Sambamba](https://github.com/biod/sambamba)
- [Freebayes](https://github.com/freebayes/freebayes)


For the GATK method:

- GATK Docker container 
- Fastp
- BWA-mem2


## Install required packages

Install packages for *freebayes* method:

	conda create -n beans -c bioconda python fastp bwa-mem2 samtools sambamba freebayes
	conda activate beans
	
Install packages for *GATK* method:

	docker pull broadinstitute/gatk
	docker run --rm -it -v `pwd`:`pwd` --workdir `pwd` broadinstitute/gatk
	
In docker container install:
	
	apt update
	apt install fastp
	curl -L https://github.com/bwa-mem2/bwa-mem2/releases/download/v2.2.1/bwa-mem2-2.2.1_x64-linux.tar.bz2 | tar jxf -
	export PATH="bwa-mem2-2.2.1_x64-linux/:$PATH"
	
Note: *\`pwd\`* is used in this context to ensure that the Docker container accesses and operates within the same working directory as the current directory of the system. This ensures that the Docker container runs in the same location as the data.


## Manual 

To execute *BEANS*, a file containing metadata of the sequencing libraries to be analyzed is required. This file must be generated by the user in TSV \(Tab-Separated Values\) format. It should consist of two columns: ID and R1.
The *ID* column will contain the names of the samples/individuals, while *R1* represents the name of the sequencing library. Example\:

| ID     | R1            |
| ------ | --------------|
| G2336  | L_G2336.fq.gz |
| G2279  | L_G2279.fq.gz |
| G16043 | L_G16043.fq.gz|
| G2335  | L_G2335.fq.gz |
| G1758  | L_G1758.fq.gz |
| G2320  | L_G2320.fq.gz |



The options are:

	beans.py [-h] [--output_dir OUTPUT_DIR]
		      [--metadata METADATA_FILE]
        	      [--skip_clean] 
        	      [--threads THREADS]
        	      [--quality QUALITY]
        	      [--genome GENOME]
        	      [--method {freebayes,gatk}]


	Arguments:
	  -h, --help            show this help message and exit
	  --output_dir OUTPUT_DIR, -o OUTPUT_DIR
        	                Output directory
	  --metadata METADATA, -m METADATA_FILE
        	                Metadata file
	  --skip_clean, -s      Skip cleaning step
	  --threads THREADS, -t THREADS
        	                Number of CPUs, default 4 (Optional)
	  --quality QUALITY, -q QUALITY
        	                Phread Quality, default Q>25 (Optional)
	  --genome GENOME, -g GENOME
        	                Reference genome in fasta format
	  --method {freebayes,gatk}, -M {freebayes,gatk}
        	                Method to use: freebayes or gatk

Tipical run:

	beans.py -o variants -m metadata.tsv -g ref.fasta -m freebayes
	
We highly recommend running *BEANS* in the same folder location as the sequencing files.



# Contact

Comments and inquiries, send a email: aragon.marco@inifap.gob.mx 



